pydoctor/astbuilder.py

```diff
@@ -904,23 +1065,38 @@ def parseFile(self, path: Path) -> Optional[ast.AST]:
 
 model.System.defaultBuilder = ASTBuilder
 
-def findAll(modast, mod):
+def findAll(mod_ast: ast.Module, mod: model.Module) -> None:
     """Find and attempt to parse into a list of names the __all__ of a module's AST."""
-    for node in modast.body:
+    for node in mod_ast.body:
         if isinstance(node, ast.Assign) and \
                len(node.targets) == 1 and \
                isinstance(node.targets[0], ast.Name) and \
                node.targets[0].id == '__all__':
             if mod.all is not None:
-                mod.system.msg('all', "multiple assignments to %s.__all__ ??"%(mod.fullName(),))
+                mod.report(
+                    'Multiple assignments to "__all__"',
+                    section='all', lineno_offset=node.lineno)
             if not isinstance(node.value, (ast.List, ast.Tuple)):
-                mod.system.msg('all', "couldn't parse %s.__all__"%(mod.fullName(),))
+                mod.report(
+                    'Cannot parse value assigned to "__all__"',
+                    section='all', lineno_offset=node.lineno)
                 continue
             items = node.value.elts
             names = []
-            for item in items:
-                if not isinstance(item, ast.Str):
-                    mod.system.msg('all', "couldn't parse %s.__all__"%(mod.fullName(),))
-                    continue
-                names.append(item.s)
+            for idx, item in enumerate(items):
+                try:
+                    name: object = ast.literal_eval(item)
+                except ValueError:
+                    mod.report(
+                        f'Cannot parse element {idx} of "__all__"',
+                        section='all', lineno_offset=node.lineno)
+                else:
+                    if isinstance(name, str):
+                        names.append(name)
+                    else:
+                        mod.report(
+                            f'Element {idx} of "__all__" has '
+                            f'type "{type(name).__name__}", expected "str"',
+                            section='all', lineno_offset=node.lineno)
+            if names or not items:
```

Comment:
`mod.all` is `None` by default, meaning that `__all__` is undefined, which for `from module import *` means that everything that doesn't start with an underscore is exported. An empty sequence means that the module explicitly chose to not export anything.

This `if` makes sure we only replace `None` with an empty sequence if we parsed an empty sequence and not if we parsed a non-empty sequence but found no names in it.

That said, I'm not sure what is the best behavior in the face of unparseable elements. I checked CPython and it doesn't accept any types other than `str`, but it does support computed strings while pydoctor only accepts literals. So the "Cannot parse" warnings probably indicate limitations in pydoctor, while the "expected 'str'" warning almost certainly indicates broken source code.

If we run into limitations of pydoctor, should we ignore the whole assignment or try to salvage as much as possible? And in the latter case, if what we can salvage is nothing, should we ignore the assignment after all or should we treat it as an empty sequence?

Reply:
> Yes. I think that is ok to include any `str` value found in `__all__` and ignore+raise-warning for computed values.
> 
> If nothing can be salvaged, treat it as empty sequence to signal that this was not `None`.
> No need for extra warning as we already have warnings raised for each member.
> 
> It looks to me like this is the current implementation.

No, the current implementation handles the case where nothing can be salvaged (only computed or non-`str` elements) by ignoring the assignment: `names` will be the empty list (nothing salvaged), while `not items` will be `False` because there were items (AST did contain items).

> See my suggestion for the code above.
> Feel free to ignore the suggestion.
> I think that makes the code easier to read (at least for myself)

I like the suggestion better as well, so I'll apply it. But I wanted to stress that this does change the behavior: in the case where nothing can be salvaged, we'll apply `__all__ = []`, instead of ignoring the assignment.